pootle/core/response.py

```diff
@@ -0,0 +1,127 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) Pootle contributors.
+#
+# This file is a part of the Pootle project. It is distributed under the GPL3
+# or later license. See the LICENSE file for a copy of the license and the
+# AUTHORS file for copyright and authorship information.
+
+import logging
+
+
+logger = logging.getLogger(__name__)
+
+
+class ItemResponse(object):
+
+    def __init__(self, response, action_type, complete=True, msg=None, **kwargs):
+        self.response = response
+        self.action_type = action_type
+        self.complete = complete
+        self.kwargs = kwargs
+        self.msg = msg or ""
+
+    def __str__(self):
+        extra = ""
+        if self.failed:
+            extra = " FAILED"
+        if self.msg:
+            extra = "%s %s" % (extra, self.msg)
+        if self.kwargs:
+            extra = "%s %s" % (extra, str(self.kwargs))
+        return (
+            "<%s(%s): %s%s>"
+            % (self.__class__.__name__,
+               self.response,
+               self.action_type,
+               extra))
+
+    @property
+    def failed(self):
+        return not self.complete
+
+
+class Response(object):
+
+    __responses__ = None
+    response_class = ItemResponse
+
+    def __init__(self, context):
+        self.context = context
+        self.__responses__ = {}
+        self.clear_cache()
+
+    def clear_cache(self):
+        for k in self.response_types:
+            self.__responses__[k] = []
+
+    def __getitem__(self, k):
+        return self.__responses__[k]
+
+    def __iter__(self):
+        for k in self.__responses__:
+            if self.__responses__[k]:
+                yield k
+
+    def __len__(self):
+        return len([x for x in self.__iter__()])
+
+    def __str__(self):
+        failed = ""
+        if self.has_failed:
+            failed = "FAIL "
+        if not self.made_changes:
+            return (
+                "<%s(%s): %sNo changes made>"
+                % (self.__class__.__name__,
+                   self.context,
+                   failed))
+        return (
+            "<%s(%s): %s%s>"
+            % (self.__class__.__name__,
+               self.context,
+               failed,
+               self.success))
+
+    @property
+    def success(self):
+        return ', '.join(
+            ["%s: %s" % (k, len(list(self.completed(k))))
+             for k in self.response_types
+             if (len(list(self.completed(k))))])
```

Comment:
Nit: is the parentheses around `len` actually necessary. It already has a lot of them.